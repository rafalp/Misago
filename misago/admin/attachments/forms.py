from django import forms
from django.utils.translation import pgettext_lazy

from ...attachments.filetypes import filetypes


def get_searchable_filetypes():
    choices = [("", pgettext_lazy("admin attachments type filter choice", "All types"))]
    choices.extend(filetypes.as_django_choices())
    return choices


class FilterAttachmentsForm(forms.Form):
    uploader = forms.CharField(
        label=pgettext_lazy("admin attachments filter form", "Uploader name contains"),
        required=False,
    )
    name = forms.CharField(
        label=pgettext_lazy("admin attachments filter form", "Name contains"),
        required=False,
    )
    filetype = forms.ChoiceField(
        label=pgettext_lazy("admin attachments filter form", "File type"),
        choices=get_searchable_filetypes,
        required=False,
    )
    status = forms.ChoiceField(
        label=pgettext_lazy("admin attachments filter form", "Status"),
        required=False,
        choices=[
            (
                "",
                pgettext_lazy(
                    "admin attachments status filter choice",
                    "All",
                ),
            ),
            (
                "posted",
                pgettext_lazy(
                    "admin attachments status filter choice",
                    "Posted",
                ),
            ),
            (
                "unused",
                pgettext_lazy(
                    "admin attachments status filter choice",
                    "Unused",
                ),
            ),
            (
                "deleted",
                pgettext_lazy(
                    "admin attachments status filter choice",
                    "Deleted",
                ),
            ),
            (
                "broken",
                pgettext_lazy(
                    "admin attachments status filter choice",
                    "Broken",
                ),
            ),
        ],
    )

    def filter_queryset(self, criteria, queryset):
        if criteria.get("uploader"):
            queryset = queryset.filter(
                uploader_slug__contains=criteria["uploader"].lower()
            )
        if criteria.get("name"):
            queryset = queryset.filter(name__icontains=criteria["name"])
        if criteria.get("filetype"):
            queryset = queryset.filter(filetype_id=criteria["filetype"])
        if criteria.get("status") == "posted":
            queryset = queryset.filter(post__isnull=False)
        elif criteria.get("status") == "unused":
            queryset = queryset.filter(post__isnull=True, is_deleted=False)
        elif criteria.get("status") == "deleted":
            queryset = queryset.filter(is_deleted=True)
        elif criteria.get("status") == "broken":
            queryset = queryset.filter(upload="")
        return queryset
