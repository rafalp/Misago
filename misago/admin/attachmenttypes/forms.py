from django import forms
from django.utils.translation import pgettext, pgettext_lazy

from ...attachments.models import AttachmentType


class AttachmentTypeForm(forms.ModelForm):
    class Meta:
        model = AttachmentType
        fields = [
            "name",
            "extensions",
            "mimetypes",
            "size_limit",
            "status",
            "limit_uploads_to",
            "limit_downloads_to",
        ]
        labels = {
            "name": pgettext_lazy("admin attachment type form", "Type name"),
            "extensions": pgettext_lazy(
                "admin attachment type form", "File extensions"
            ),
            "mimetypes": pgettext_lazy("admin attachment type form", "Mimetypes"),
            "size_limit": pgettext_lazy(
                "admin attachment type form", "Maximum allowed uploaded file size"
            ),
            "status": pgettext_lazy("admin attachment type form", "Status"),
            "limit_uploads_to": pgettext_lazy(
                "admin attachment type form", "Limit uploads to"
            ),
            "limit_downloads_to": pgettext_lazy(
                "admin attachment type form", "Limit downloads to"
            ),
        }
        help_texts = {
            "extensions": pgettext_lazy(
                "admin attachment type form",
                "List of comma separated file extensions associated with this attachment type.",
            ),
            "mimetypes": pgettext_lazy(
                "admin attachment type form",
                "Optional list of comma separated mime types associated with this attachment type.",
            ),
            "size_limit": pgettext_lazy(
                "admin attachment type form",
                "Maximum allowed uploaded file size for this type, in kb. This setting is deprecated and has no effect. It will be deleted in Misago 1.0.",
            ),
            "status": pgettext_lazy(
                "admin attachment type form",
                "Controls this attachment type availability on your site.",
            ),
            "limit_uploads_to": pgettext_lazy(
                "admin attachment type form",
                "If you wish to limit option to upload files of this type to users with specific roles, select them on this list. Otherwise don't select any roles to allow all users with permission to upload attachments to be able to upload attachments of this type.",
            ),
            "limit_downloads_to": pgettext_lazy(
                "admin attachment type form",
                "If you wish to limit option to download files of this type to users with specific roles, select them on this list. Otherwise don't select any roles to allow all users with permission to download attachments to be able to download attachments of this type.",
            ),
        }
        widgets = {
            "limit_uploads_to": forms.CheckboxSelectMultiple,
            "limit_downloads_to": forms.CheckboxSelectMultiple,
        }

    def clean_extensions(self):
        data = self.clean_list(self.cleaned_data["extensions"])
        if not data:
            raise forms.ValidationError(
                pgettext("admin attachment type form", "This field is required.")
            )
        return data

    def clean_mimetypes(self):
        data = self.cleaned_data["mimetypes"]
        if data:
            return self.clean_list(data)

    def clean_list(self, value):
        items = [v.lstrip(".") for v in value.lower().replace(" ", "").split(",")]
        return ",".join(set(filter(bool, items)))
