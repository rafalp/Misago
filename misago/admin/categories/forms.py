from django import forms
from django.db import models
from django.utils.html import conditional_escape, mark_safe
from django.utils.translation import pgettext_lazy
from mptt.forms import TreeNodeChoiceField, TreeNodeMultipleChoiceField

from ...admin.forms import YesNoSwitch
from ...categories import THREADS_ROOT_NAME
from ...categories.models import Category
from ...core.validators import validate_color_hex, validate_sluggable
from ...threads.threadtypes import trees_map


class AdminCategoryFieldMixin:
    def __init__(self, *args, **kwargs):
        self.base_level = kwargs.pop("base_level", 1)
        kwargs["level_indicator"] = kwargs.get("level_indicator", "- - ")

        threads_tree_id = trees_map.get_tree_id_for_root(THREADS_ROOT_NAME)
        queryset = Category.objects.filter(tree_id=threads_tree_id)
        if not kwargs.pop("include_root", False):
            queryset = queryset.exclude(special_role="root_category")

        kwargs.setdefault("queryset", queryset)

        super().__init__(*args, **kwargs)

    def _get_level_indicator(self, obj):
        level = getattr(obj, obj._mptt_meta.level_attr) - self.base_level
        if level > 0:
            return mark_safe(conditional_escape(self.level_indicator) * level)
        return ""


class AdminCategoryChoiceField(AdminCategoryFieldMixin, TreeNodeChoiceField):
    pass


class AdminCategoryMultipleChoiceField(
    AdminCategoryFieldMixin, TreeNodeMultipleChoiceField
):
    pass


class CategoryForm(forms.ModelForm):
    name = forms.CharField(
        label=pgettext_lazy("admin category form", "Name"),
        validators=[validate_sluggable()],
    )
    short_name = forms.CharField(
        label=pgettext_lazy("admin category form", "Short name"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Optional, alternative or abbreviated name (eg. abbreviation) used on threads list.",
        ),
        validators=[validate_sluggable()],
    )
    color = forms.CharField(
        label=pgettext_lazy("admin category form", "Color"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Optional but recommended, should be in hex format, eg. #F5A9B8.",
        ),
        validators=[validate_color_hex],
    )
    description = forms.CharField(
        label=pgettext_lazy("admin category form", "Description"),
        max_length=2048,
        required=False,
        widget=forms.Textarea(attrs={"rows": 3}),
        help_text=pgettext_lazy(
            "admin category form",
            "Optional description explaining category intented purpose.",
        ),
    )
    css_class = forms.CharField(
        label=pgettext_lazy("admin category form", "CSS class"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Optional CSS class used to customize this category's appearance from themes.",
        ),
    )
    allow_polls = YesNoSwitch(
        label=pgettext_lazy("admin category form", "Allow polls"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Controls if users can start new polls in this category.",
        ),
    )
    allow_list_access = YesNoSwitch(
        label=pgettext_lazy(
            "admin category form",
            'Allow users without the "browse contents" permission to access the threads list',
        ),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            'Enabling this option will allow users with permission to "see" this category to access it\'s threads list page. They will receive an error if they try to see threads replies.',
        ),
    )
    limit_threads_visibility = YesNoSwitch(
        label=pgettext_lazy(
            "admin category form", "Show users only threads that they started"
        ),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Enabling this option will limit users access to threads in this category to only the threads they have started. Moderators will still have access to all threads.",
        ),
    )
    is_closed = YesNoSwitch(
        label=pgettext_lazy("admin category form", "Closed category"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Only members with valid permissions can post in closed categories.",
        ),
    )
    require_threads_approval = YesNoSwitch(
        label=pgettext_lazy("admin category form", "Threads"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "All threads started in this category will require moderator approval.",
        ),
    )
    require_replies_approval = YesNoSwitch(
        label=pgettext_lazy("admin category form", "Replies"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "All replies posted in this category will require moderator approval.",
        ),
    )
    require_edits_approval = YesNoSwitch(
        label=pgettext_lazy("admin category form", "Edits"),
        required=False,
        help_text=pgettext_lazy(
            "admin category form",
            "Will make all edited replies return to unapproved state for moderator to review.",
        ),
    )
    prune_started_after = forms.IntegerField(
        label=pgettext_lazy("admin category form", "Thread age"),
        min_value=0,
        help_text=pgettext_lazy(
            "admin category form",
            "Prune thread if number of days since its creation is greater than specified. Enter 0 to disable this pruning criteria.",
        ),
    )
    prune_replied_after = forms.IntegerField(
        label=pgettext_lazy("admin category form", "Last reply"),
        min_value=0,
        help_text=pgettext_lazy(
            "admin category form",
            "Prune thread if number of days since last reply is greater than specified. Enter 0 to disable this pruning criteria.",
        ),
    )

    class Meta:
        model = Category
        fields = [
            "name",
            "short_name",
            "color",
            "description",
            "css_class",
            "allow_polls",
            "allow_list_access",
            "limit_threads_visibility",
            "is_closed",
            "require_threads_approval",
            "require_replies_approval",
            "require_edits_approval",
            "prune_started_after",
            "prune_replied_after",
            "archive_pruned_in",
        ]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.setup_new_parent_field()
        self.setup_copy_permissions_field()
        self.setup_archive_pruned_in_field()

    def setup_new_parent_field(self):
        parent_queryset = Category.objects.all_categories(True).order_by("lft")
        if self.instance.pk:
            not_siblings = models.Q(lft__lt=self.instance.lft)
            not_siblings = not_siblings | models.Q(rght__gt=self.instance.rght)
            parent_queryset = parent_queryset.filter(not_siblings)

        self.fields["new_parent"] = AdminCategoryChoiceField(
            label=pgettext_lazy("admin category form", "Parent category"),
            queryset=parent_queryset,
            initial=self.instance.parent,
            empty_label=None,
        )

    def setup_copy_permissions_field(self):
        self.fields["copy_permissions"] = AdminCategoryChoiceField(
            label=pgettext_lazy("admin category form", "Copy permissions"),
            help_text=pgettext_lazy(
                "admin category form",
                "You can replace this category permissions with permissions copied from the category selected here.",
            ),
            queryset=Category.objects.all_categories(),
            empty_label=pgettext_lazy("admin category form", "Don't copy permissions"),
            required=False,
        )

    def setup_archive_pruned_in_field(self):
        self.fields["archive_pruned_in"] = AdminCategoryChoiceField(
            label=pgettext_lazy("admin category form", "Archive"),
            help_text=pgettext_lazy(
                "admin category form",
                "Instead of being deleted, pruned threads can be moved to designated category.",
            ),
            queryset=Category.objects.all_categories(),
            empty_label=pgettext_lazy(
                "admin category form", "Don't archive pruned threads"
            ),
            required=False,
        )

    def clean_copy_permissions(self):
        data = self.cleaned_data["copy_permissions"]
        if data and data.pk == self.instance.pk:
            message = pgettext_lazy(
                "admin category form",
                "Permissions cannot be copied from category into itself.",
            )
            raise forms.ValidationError(message)
        return data

    def clean_archive_pruned_in(self):
        data = self.cleaned_data["archive_pruned_in"]
        if data and data.pk == self.instance.pk:
            message = pgettext_lazy(
                "admin category form", "Category cannot act as archive for itself."
            )
            raise forms.ValidationError(message)
        return data

    def clean(self):
        data = super().clean()
        self.instance.set_name(data.get("name"))
        return data


class DeleteCategoryForm(forms.ModelForm):
    class Meta:
        model = Category
        fields = []

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.setup_fields()

    def setup_fields(self):
        content_queryset = Category.objects.all_categories().order_by("lft")
        self.fields["move_threads_to"] = AdminCategoryChoiceField(
            label=pgettext_lazy("admin category form", "Move category threads to"),
            queryset=content_queryset,
            initial=self.instance.parent,
            empty_label=pgettext_lazy("admin category form", "Delete with category"),
            required=False,
        )

        not_siblings = models.Q(lft__lt=self.instance.lft)
        not_siblings = not_siblings | models.Q(rght__gt=self.instance.rght)
        children_queryset = Category.objects.all_categories(True)
        children_queryset = children_queryset.filter(not_siblings).order_by("lft")

        if children_queryset.exists():
            self.fields["move_children_to"] = AdminCategoryChoiceField(
                label=pgettext_lazy("admin category form", "Move child categories to"),
                queryset=children_queryset,
                empty_label=pgettext_lazy(
                    "admin category form", "Delete with category"
                ),
                required=False,
            )

    def clean(self):
        data = super().clean()

        if data.get("move_threads_to"):
            if data["move_threads_to"].pk == self.instance.pk:
                message = pgettext_lazy(
                    "admin category form",
                    "You are trying to move this category threads to itself.",
                )
                raise forms.ValidationError(message)

            moving_to_child = self.instance.has_child(data["move_threads_to"])
            if moving_to_child and not data.get("move_children_to"):
                message = pgettext_lazy(
                    "admin category form",
                    "You are trying to move this category threads to a child category that will also be deleted.",
                )
                raise forms.ValidationError(message)

        return data
