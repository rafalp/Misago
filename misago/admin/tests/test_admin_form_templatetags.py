from unittest.mock import Mock

import pytest
from django import forms
from django.template import Context, Template, TemplateSyntaxError

from ..forms import YesNoSwitch
from ..templatetags.misago_admin_form import (
    get_field_image_dimensions,
    is_multiple_choice_field,
    is_radio_select_field,
    is_select_field,
    is_textarea_field,
    is_yesno_switch_field,
    render_attrs,
    render_bool_attrs,
)


class Form(forms.Form):
    text_field = forms.CharField(
        label="Hello!", max_length=255, help_text="I am a help text."
    )
    textarea_field = forms.CharField(
        label="Message", max_length=255, widget=forms.Textarea()
    )
    select_field = forms.ChoiceField(
        label="Choice", choices=(("y", "Yes"), ("n", "No"))
    )
    ratio_select_field = forms.ChoiceField(
        label="Choice",
        choices=(("y", "Yes"), ("n", "No")),
        widget=forms.RadioSelect,
    )
    checkbox_select_field = forms.MultipleChoiceField(
        label="Color",
        choices=(("r", "Red"), ("g", "Green"), ("b", "Blue")),
        widget=forms.CheckboxSelectMultiple,
    )
    multiple_select_field = forms.MultipleChoiceField(
        label="Rank", choices=(("r", "Red"), ("g", "Green"), ("b", "Blue"))
    )
    yesno_field = YesNoSwitch(label="Switch")
    image_field = forms.ImageField(label="Image!", help_text="I am a help text.")


@pytest.fixture
def form():
    return Form()


def render(template_str, form=None):
    base_template = "{%% load misago_admin_form %%} %s"
    context = Context({"form": form or Form()})
    template = Template(base_template % template_str)
    return template.render(context).strip()


def test_row_with_field_input_is_rendered():
    html = render("{% form_row form.text_field %}")
    assert "id_text_field" in html


def test_row_with_field_input_and_label_css_class_is_rendered():
    html = render('{% form_row form.text_field label_class="col-md-3" %}')
    assert "id_text_field" in html
    assert "col-md-3" in html


def test_row_with_field_input_and_field_css_class_is_rendered():
    html = render('{% form_row form.text_field field_class="col-md-9" %}')
    assert "id_text_field" in html
    assert "col-md-9" in html


def test_row_with_field_input_and_label_and_field_css_classes_is_rendered():
    html = render('{% form_row form.text_field "col-md-3" "col-md-9" %}')
    assert "id_text_field" in html
    assert "col-md-3" in html
    assert "col-md-9" in html


def test_field_label_is_rendered():
    html = render("{% form_row form.text_field %}")
    assert "Hello!" in html


def test_field_help_text_is_rendered():
    html = render("{% form_row form.text_field %}")
    assert "I am a help text." in html


def test_tag_without_field_raises_exception():
    with pytest.raises(TemplateSyntaxError):
        render("{% form_row %}")


def test_image_row_with_field_input_is_rendered():
    html = render("{% form_image_row form.image_field %}")
    assert "id_image_field" in html


def test_image_row_with_field_input_and_label_css_class_is_rendered():
    html = render('{% form_image_row form.image_field label_class="col-md-3" %}')
    assert "id_image_field" in html
    assert "col-md-3" in html


def test_image_row_with_field_input_and_field_css_class_is_rendered():
    html = render('{% form_image_row form.image_field field_class="col-md-9" %}')
    assert "id_image_field" in html
    assert "col-md-9" in html


def test_image_row_with_field_input_and_label_and_field_css_classes_is_rendered():
    html = render('{% form_image_row form.image_field "col-md-3" "col-md-9" %}')
    assert "id_image_field" in html
    assert "col-md-3" in html
    assert "col-md-9" in html


def test_image_field_label_is_rendered():
    html = render("{% form_image_row form.image_field %}")
    assert "Image!" in html


def test_image_field_help_text_is_rendered():
    html = render("{% form_image_row form.image_field %}")
    assert "I am a help text." in html


class TestImage:
    url = "test-image.png"

    def __str__(self):
        return self.url


def render_image(template_str):
    base_template = "{%% load misago_admin_form %%} %s"
    context = Context({"form": Form(initial={"image_field": TestImage()})})
    template = Template(base_template % template_str)
    return template.render(context).strip()


def test_image_row_with_value_renders_label(snapshot):
    html = render_image("{% form_image_row form.image_field %}")
    assert snapshot == html


def test_image_row_with_value_renders_help_text(snapshot):
    html = render_image("{% form_image_row form.image_field %}")
    assert snapshot == html


def test_image_row_with_value_renders_image_preview(snapshot):
    html = render_image("{% form_image_row form.image_field %}")
    assert ('src="%s"' % TestImage.url) in html
    assert snapshot == html


def test_image_row_with_value_renders_input(snapshot):
    html = render_image("{% form_image_row form.image_field %}")
    assert "id_image_field" in html
    assert snapshot == html


def test_image_row_with_value_renders_input_and_label_css_class(snapshot):
    html = render_image('{% form_image_row form.image_field label_class="col-md-3" %}')
    assert "id_image_field" in html
    assert "col-md-3" in html
    assert snapshot == html


def test_image_row_with_value_renders_input_and_field_css_class(snapshot):
    html = render_image('{% form_image_row form.image_field field_class="col-md-9" %}')
    assert "id_image_field" in html
    assert "col-md-9" in html
    assert snapshot == html


def test_image_row_with_value_renders_input_and_label_and_field_css_classes(snapshot):
    html = render_image('{% form_image_row form.image_field "col-md-3" "col-md-9" %}')
    assert "id_image_field" in html
    assert "col-md-3" in html
    assert "col-md-9" in html
    assert snapshot == html


def test_image_tag_without_field_raises_exception():
    with pytest.raises(TemplateSyntaxError):
        render("{% form_image_row %}")


def test_for_field_with_radio_select_widget_filter_returns_true(form):
    assert is_radio_select_field(form["ratio_select_field"])


def test_for_field_without_radio_select_widget_filter_returns_false(form):
    assert not is_radio_select_field(form["text_field"])


def test_row_with_radio_select_field_is_rendered(snapshot):
    html = render_image("{% form_image_row form.ratio_select_field %}")
    assert snapshot == html


def test_for_field_with_select_widget_filter_returns_true(form):
    assert is_select_field(form["select_field"])


def teste_for_field_without_select_widget_filter_returns_false(form):
    assert not is_select_field(form["text_field"])


def test_row_with_select_field_is_rendered(snapshot):
    html = render_image("{% form_image_row form.select_field %}")
    assert snapshot == html


def test_for_field_with_checkbox_select_widget_filter_returns_true(form):
    assert is_multiple_choice_field(form["checkbox_select_field"])


def test_for_field_without_checkbox_select_widget_filter_returns_false(form):
    assert not is_multiple_choice_field(form["text_field"])


def test_row_with_checkbox_select_field_is_rendered(snapshot):
    html = render_image("{% form_image_row form.checkbox_select_field %}")
    assert snapshot == html


def test_for_field_with_multiple_select_widget_filter_returns_true(form):
    assert is_multiple_choice_field(form["multiple_select_field"])


def test_for_field_without_multiple_select_widget_filter_returns_false(form):
    assert not is_multiple_choice_field(form["text_field"])


def test_row_with_select_multiple_field_is_rendered(snapshot):
    html = render_image("{% form_image_row form.multiple_select_field %}")
    assert snapshot == html


def test_for_field_with_textarea_widget_filter_returns_true(form):
    assert is_textarea_field(form["textarea_field"])


def test_for_field_without_textarea_widget_filter_returns_false(form):
    assert not is_textarea_field(form["text_field"])


def test_row_with_textarea_field_is_rendered(snapshot):
    html = render_image("{% form_image_row form.textarea_field %}")
    assert snapshot == html


def test_for_yes_no_field_filter_returns_true(form):
    assert is_yesno_switch_field(form["yesno_field"])


def test_for_non_yes_no_field_filter_returns_false(form):
    assert not is_yesno_switch_field(form["text_field"])


def test_row_with_yes_no_field_is_rendered(snapshot):
    html = render_image("{% form_image_row form.yesno_field %}")
    assert snapshot == html


def test_specified_class_name_is_rendered():
    result = render_attrs({"attrs": {}}, class_name="form-control")
    assert result == 'class="form-control"'


def test_specified_class_name_overrided_by_class_attr():
    result = render_attrs({"attrs": {"class": "custom"}}, class_name="form-control")
    assert result == 'class="custom"'


def test_attr_with_string_value_is_rendered():
    result = render_attrs({"attrs": {"name": "lorem"}})
    assert result == 'name="lorem"'


def test_attr_with_int_value_is_rendered():
    result = render_attrs({"attrs": {"cols": 5}})
    assert result == 'cols="5"'


def test_attr_with_boolean_true_value_is_not_rendered():
    result = render_attrs({"attrs": {"selected": True}})
    assert result == ""


def test_attr_with_boolean_false_value_is_not_rendered():
    result = render_attrs({"attrs": {"selected": False}})
    assert result == ""


def test_attr_with_none_value_is_not_rendered():
    result = render_attrs({"attrs": {"selected": None}})
    assert result == ""


def test_attr_name_is_escaped():
    result = render_attrs({"attrs": {'"': "test"}})
    assert result == '&quot;="test"'


def test_attr_value_is_escaped():
    result = render_attrs({"attrs": {"name": '"'}})
    assert result == 'name="&quot;"'


def test_multiple_valid_attrs_are_rendered():
    result = render_attrs({"attrs": {"name": "lorem", "cols": 5}})
    assert result == 'name="lorem" cols="5"'


def test_empty_attr_dict_is_not_rendered():
    result = render_attrs({"attrs": {}})
    assert result == ""


def test_attr_with_boolean_true_value_is_rendered():
    result = render_bool_attrs({"bool": True})
    assert result == "bool"


def test_bool_attr_with_string_value_is_not_rendered():
    result = render_bool_attrs({"name": "hello"})
    assert result == ""


def test_bool_attr_with_int_value_is_not_rendered():
    result = render_bool_attrs({"col": 13})
    assert result == ""


def test_bool_attr_with_boolean_false_value_is_not_rendered():
    result = render_bool_attrs({"selected": False})
    assert result == ""


def test_bool_attr_with_none_value_is_not_rendered():
    result = render_bool_attrs({"selected": None})
    assert result == ""


def test_bool_attr_with_false_int_value_is_not_rendered():
    result = render_bool_attrs({"selected": 0})
    assert result == ""


def test_multiple_bool_attrs_with_boolean_true_value_are_rendered():
    result = render_bool_attrs({"selected": True, "required": True})
    assert result == "selected required"


def test_only_bool_attrs_with_boolean_true_value_are_rendered():
    result = render_bool_attrs({"bool": True, "string": "hello", "int": 123})
    assert result == "bool"


def test_util_turns_dimensions_tuple_into_dict():
    assert get_field_image_dimensions((20, 40)) == {"width": 20, "height": 40}


def test_util_returs_none_if_dimensions_are_empty():
    assert get_field_image_dimensions(None) is None
