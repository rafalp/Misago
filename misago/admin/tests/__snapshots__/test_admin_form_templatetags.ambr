# serializer version: 1
# name: test_image_row_with_value_renders_help_text
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label">
        Image!:
      </label>
      <div class="">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_image_row_with_value_renders_image_preview
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label">
        Image!:
      </label>
      <div class="">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_image_row_with_value_renders_input
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label">
        Image!:
      </label>
      <div class="">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_image_row_with_value_renders_input_and_field_css_class
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label">
        Image!:
      </label>
      <div class="col-md-9">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_image_row_with_value_renders_input_and_label_and_field_css_classes
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label col-md-3">
        Image!:
      </label>
      <div class="col-md-9">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_image_row_with_value_renders_input_and_label_css_class
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label col-md-3">
        Image!:
      </label>
      <div class="">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_image_row_with_value_renders_label
  '''
  <div class="form-group">
      <label for="id_image_field" class="control-label">
        Image!:
      </label>
      <div class="">
        <div class="control-image-preview">
          <img src="test-image.png" alt="" />
          
        </div>
        
  
    <input accept="image/*" class="form-control" id="id_image_field" name="image_field" type="file" value="test-image.png" required is_initial >
  
        
        
          <small class="form-text text-muted">I am a help text.</small>
        
      </div>
    </div>
  '''
# ---
# name: test_row_with_checkbox_select_field_is_rendered
  '''
  <div class="form-group">
    <label for="" class="control-label">
      Color:
    </label>
    <div class="controls control-radioselect controls control-checkboxselect ">
      
  
    
  
    <div class="form-check">
      <input class_name="form-check-input" id="id_checkbox_select_field_0" name="checkbox_select_field" type="checkbox" value="r">
      <label class="form-check-label" for="id_checkbox_select_field_0">
        Red
      </label>
    </div>
  
    <div class="form-check">
      <input class_name="form-check-input" id="id_checkbox_select_field_1" name="checkbox_select_field" type="checkbox" value="g">
      <label class="form-check-label" for="id_checkbox_select_field_1">
        Green
      </label>
    </div>
  
    <div class="form-check">
      <input class_name="form-check-input" id="id_checkbox_select_field_2" name="checkbox_select_field" type="checkbox" value="b">
      <label class="form-check-label" for="id_checkbox_select_field_2">
        Blue
      </label>
    </div>
  
  
      
      
    </div>
  </div>
  '''
# ---
# name: test_row_with_radio_select_field_is_rendered
  '''
  <div class="form-group">
    <label for="" class="control-label">
      Choice:
    </label>
    <div class="controls control-radioselect ">
      
  
    
  
    
      <div class="form-check">
        <input class_name="form-check-input" id="id_ratio_select_field_0" name="ratio_select_field" type="radio" value="y">
        <label class="form-check-label" for="id_ratio_select_field_0">
          Yes
        </label>
      </div>
    
      <div class="form-check">
        <input class_name="form-check-input" id="id_ratio_select_field_1" name="ratio_select_field" type="radio" value="n">
        <label class="form-check-label" for="id_ratio_select_field_1">
          No
        </label>
      </div>
    
  
  
      
      
    </div>
  </div>
  '''
# ---
# name: test_row_with_select_field_is_rendered
  '''
  <div class="form-group">
    <label for="id_select_field" class="control-label">
      Choice:
    </label>
    <div class="">
      
  
    
  <select class="form-control" id="id_select_field" name="select_field" required>
    
      <option value="y" >
        Yes
      </option>
    
      <option value="n" >
        No
      </option>
    
  </select>
  
      
      
    </div>
  </div>
  '''
# ---
# name: test_row_with_select_multiple_field_is_rendered
  '''
  <div class="form-group">
    <label for="id_multiple_select_field" class="control-label">
      Rank:
    </label>
    <div class="controls control-checkboxselect ">
      
  
    
  
    <div class="form-check">
      <input class_name="form-check-input" id="" name="multiple_select_field" type="checkbox" value="r">
      <label class="form-check-label" for="">
        Red
      </label>
    </div>
  
    <div class="form-check">
      <input class_name="form-check-input" id="" name="multiple_select_field" type="checkbox" value="g">
      <label class="form-check-label" for="">
        Green
      </label>
    </div>
  
    <div class="form-check">
      <input class_name="form-check-input" id="" name="multiple_select_field" type="checkbox" value="b">
      <label class="form-check-label" for="">
        Blue
      </label>
    </div>
  
  
      
      
    </div>
  </div>
  '''
# ---
# name: test_row_with_textarea_field_is_rendered
  '''
  <div class="form-group">
    <label for="id_textarea_field" class="control-label">
      Message:
    </label>
    <div class="">
      
  
    <textarea cols="40" rows="10" maxlength="255" class="form-control" id="id_textarea_field" name="textarea_field" required ></textarea>
  
      
      
    </div>
  </div>
  '''
# ---
# name: test_row_with_yes_no_field_is_rendered
  '''
  <div class="form-group">
    <label for="" class="control-label">
      Switch:
    </label>
    <div class="controls control-yesno-switch ">
      
  
    
  
    
      <label class="radio">
        <input class="yesno-switch" id="id_yesno_field_0" name="yesno_field" type="radio" value="1">
        Yes
      </label>
    
      <label class="radio">
        <input class="yesno-switch" id="id_yesno_field_1" name="yesno_field" type="radio" value="0" checked>
        No
      </label>
    
  
  
      
      
    </div>
  </div>
  '''
# ---
